### PR TITLE
Improve Unclear error FS3204 Multicase union struct

### PR DESCRIPTION
**Pull request by edgarfgp**
Fixes https://github.com/dotnet/fsharp/issues/3648

**Problem**
- Existing error is not necessarily wrong. But should be optimized for a person who did not see it before, and maybe even for a person who might not know that fields in DUs can have optimal names.
- The implementation uses the **type** range and shows only one error that covers all the errors in Multicase DU

**Poposal** 
- Update the error message to: 
    **If a union type has more than one case and is a struct, then all fields within the union type must be given unique field names. For example: 'type A = B of b: int | C of c: int' (unique field names 'b' and 'c' assigned).** based @T-Gro suggestion

- Use the field range to report the error in all the fields that does not have a compiler autogenerated name "Item"

Update : Would be good to put his in a preview lang flag

### CODE REVIEW

**Comment by edgarfgp** [comment]
I will be adding more unit tests tomorrow ;)

**Comment by edgarfgp** [comment]
There are some baseline tests failing. I will wait for feedback about this proposal before I update all :)

File: tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        """
+        |> compile
+        |> shouldFail
+        |> withDiagnostics [
+            (Error 954, Line 4, Col 6, Line 4, Col 17, "This type definition involves an immediate cyclic reference through a struct field or inheritance relation")
+        ]
+        
+    [<Fact>]
+    let ``If a union type has more than one case and is a struct, field must be given unique name 14`` () =
+        Fsx """
```
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs:227]: This thing works as of now (since there is Item1 generated, but it does not conflict).
We should keep it working (backwards compat).

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AbEAzAzmgExAGoAfAOwEMBbGXAB0rBgAIAxCCAWACgBtADwBlAC5QArmBEA+ALq8RAT3qsAchBEBBACqxWAXl4tjLUiwCi1ekqMmzAYQAWASwwFY5FhGwtcYySIsABTO5IEAVL5ioQDmAJRAA=

File: tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+            (Error 3204, Line 6, Col 12, Line 6, Col 18, "If a union type has more than one case and is a struct, then all fields within the union type must be given unique field names. For example: 'type A = B of b: int | C of c: int' (unique field names 'b' and 'c' assigned).")
+            (Error 3204, Line 7, Col 12, Line 7, Col 18, "If a union type has more than one case and is a struct, then all fields within the union type must be given unique field names. For example: 'type A = B of b: int | C of c: int' (unique field names 'b' and 'c' assigned).")
+        ]
+        
+    [<Fact>]
+    let ``If a union type has more than one case and is a struct, field must be given unique name 10`` () =
+        Fsx """
+namespace Foo
+[<Struct>]
+type StructUnion =
```
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs:147]: This should keep working as it is now.
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AbEAzAzmgExAGoAfAOwEMBbGXAB0rBgAIAxCCAWACgBtADwBlAC5QArmBEA+ALq8RAT3qtREqQFVyASwjkWAXl4sTLUiwCCLCNhaUQLbeREsAVHYCMD3GKcBzY1NzACFrW1AWHyh/QJNzAGEwyN9yPyA==

Generated field a,a1,b,item.
Since there is a single "item", it works => keep it working for backwards compat.

File: tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs [hunk 3]
```diff
@@ -0,0 +1,250 @@
+﻿namespace FSharp.Compiler.ComponentTests.Conformance
+
+open Xunit
+open FSharp.Test.Compiler
+
+module MultiCaseUnionStructTypes =
```
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs:6]: Please add a test for field names assigning, but having conflicts.
1. One testconflicting names within one case
2. second test conflicting names across cases
3. third test a conflict between user-provided name called "item" and the auto-generated one called "item" => that one actually works now, because compiler puts in an underscore: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AbEAzAzmgExAGoAfAOwEMBbGXAB0rBgAIAxCCAWACgBtADwBlAC5QArmBEA+ALq8RAT3qtREqQFVyASwjkWAXl4sTLUiwCCLCNhPbyI46fMAha7e0iY1EPZFA=

**Review by T-Gro — CHANGES_REQUESTED** [review]

**Comment by T-Gro** [comment]
Good to see this, thank you @edgarfgp . Added a few comments about scenarios where I would like to maintain backwards compat. with what was compiling until now.

( We do not have to be strict about code that has errors even before. But for working code, I would prefer not to introduce breaking changes unless we have to)

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 4]
```diff
@@ -3256,9 +3256,10 @@ module EstablishTypeDefinitionCores =
                     let hasRQAAttribute = HasFSharpAttribute cenv.g cenv.g.attrib_RequireQualifiedAccessAttribute tycon.Attribs
                     let unionCases = TcRecdUnionAndEnumDeclarations.TcUnionCaseDecls cenv envinner innerParent thisTy thisTyInst hasRQAAttribute tpenv unionCases
                     if tycon.IsStructRecordOrUnionTycon && unionCases.Length > 1 then 
-                      let fieldNames = [ for uc in unionCases do for ft in uc.FieldTable.TrueInstanceFieldsAsList do yield ft.LogicalName ]
-                      if fieldNames |> List.distinct |> List.length <> fieldNames.Length then 
-                          errorR(Error(FSComp.SR.tcStructUnionMultiCaseDistinctFields(), m))
+                      let fieldNames = [ for uc in unionCases do for ft in uc.FieldTable.TrueInstanceFieldsAsList do yield ft ]
+                      for field in fieldNames do
```
[the remark this answers is not in the document]
Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs]: Yeah, my first attempt was trying to simplify / make it explicit the errors in this case. But this is difficult to get right. Now I see why these only reports the error at the type and not the fields :(. I have updated to not broke existing working code

File: tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs (hunk 1 above)
Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs:638]: Updated .

File: tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs (hunk 2 above)
Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs:567]: Updated

File: tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs (hunk 3 above)
Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs:6]: I have added unit tests for this I believe :)

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 5]
```diff
@@ -3162,8 +3162,21 @@ module EstablishTypeDefinitionCores =
...
+            
+            let multiCaseUnionStructCheck (unionCases: UnionCase list) =
+                if tycon.IsStructRecordOrUnionTycon && unionCases.Length > 1 then 
+                    let fieldNames = [ for uc in unionCases do for ft in uc.FieldTable.TrueInstanceFieldsAsList do yield ft ]
+                    let fieldNames = fieldNames |> List.map(fun x -> (x.LogicalName, x.rfield_name_generated, x.Range))
+
+                    if fieldNames |> List.distinctBy(fun (name, _, _) -> name) |> List.length <> fieldNames.Length then
+                        let fieldRanges =
+                            fieldNames
+                            |> List.filter(fun (name, isGenerated, _) -> (not isGenerated || name = "Item") || isGenerated)
```
Comment (0101) [inline src/Compiler/Checking/CheckDeclarations.fs:3175]: This will always be `true` so it's either redundant or needs to be fixed.

Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs:3175]: @0101 
This is to cover this scenario. 
-  | A of Item: int  -> Item is not generated(manually added by user)
-  | B of string  -> compiler autogenerates Item as a LogicalName so we need to report it also


```fsharp

     [<Fact>]
     let ``If a union type has more than one case and is a struct, field must be given unique name 15`` () =
         Fsx """
 namespace Foo
 [<Struct>]
 type StructUnion =
     | A of Item: int
     | B of string
         """
         |> compile
         |> shouldFail
         |> withDiagnostics [
             (Error 3204, Line 5, Col 12, Line 5, Col 16, "If a union type has more than one case and is a struct, then all fields within the union type must be given unique field names. For example: 'type A = B of b: int | C of c: int' (unique field names 'b' and 'c' assigned).")
             (Error 3204, Line 6, Col 12, Line 6, Col 18, "If a union type has more than one case and is a struct, then all fields within the union type must be given unique field names. For example: 'type A = B of b: int | C of c: int' (unique field names 'b' and 'c' assigned).")
         ]
```

Comment (0101) [reply src/Compiler/Checking/CheckDeclarations.fs:3175]: @edgarfgp in that case you probably want:
```fsharp
(not isGenerated && name = "Item") || isGenerated
```
or simpler:
```fsharp
isGenerated || name = "Item"
```

But since all the tests passed anyway - what would be the case which this is supposed to filter out?

File: src/Compiler/FSComp.txt [hunk 6]
```diff
@@ -1372,7 +1372,7 @@ tcTupleStructMismatch,"One tuple type is a struct tuple, the other is a referenc
 3201,tcModuleAbbrevFirstInMutRec,"In a recursive declaration group, module abbreviations must come after all 'open' declarations and before other declarations"
 3202,tcUnsupportedMutRecDecl,"This declaration is not supported in recursive declaration groups"
 3203,parsInvalidUseOfRec,"Invalid use of 'rec' keyword"
-3204,tcStructUnionMultiCaseDistinctFields,"If a union type has more than one case and is a struct, then all fields within the union type must be given unique names."
+3204,tcStructUnionMultiCaseDistinctFields,"If a union type has more than one case and is a struct, then all fields within the union type must be given unique field names. For example: 'type A = B of b: int | C of c: int' (unique field names 'b' and 'c' assigned)."
```
Comment (psfinaki) [inline src/Compiler/FSComp.txt:1375]: @T-Gro, @edgarfgp, a bit late to the party, but I'd rather formulate it like this:
> If a multicase union type is a struct, then all union cases must have unique names. For example: 'type A = B of b: int | C of c: int'.`

Reasoning: 
1) Brevity and the accent on the fact it's valid only for structs
2) `must be given` - technically not entirely correct, they just should end up unique in the end, as `type A = | B of b: int | C of int` also works. Therefore `must have`

Also, we could technically point people to the docs, it's well explained there: https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/discriminated-unions#struct-discriminated-unions

File: src/Compiler/Checking/CheckDeclarations.fs (hunk 5 above)
Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs:3175]: Yeah, I think I was trying to be too clever :) and show only the error on ranges that not auto generated. But I think it is safer to remove the filter and just show the error on the ranges that are in the list.

Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs]: @0101 remove the need of filtering. Thanks for the review

File: tests/fsharp/typecheck/sigs/neg95.fs [hunk 7]
```diff
@@ -50,12 +50,3 @@ namespace Neg95B
 
     [<Struct>]
     type StructUnion = StructUnion of float * StructUnion 
-
-    [<Struct>]
```
Comment (psfinaki) [inline tests/fsharp/typecheck/sigs/neg95.fs:54]: Just to understand, why is this removed?

File: tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+            (Error 3204, Line 5, Col 21, Line 5, Col 27, "If a multicase union type is a struct, then all union cases must have unique names. For example: 'type A = B of b: int | C of c: int'.")
+            (Error 3204, Line 6, Col 12, Line 6, Col 18, "If a multicase union type is a struct, then all union cases must have unique names. For example: 'type A = B of b: int | C of c: int'.")
+            (Error 3204, Line 6, Col 21, Line 6, Col 22, "If a multicase union type is a struct, then all union cases must have unique names. For example: 'type A = B of b: int | C of c: int'.")
+            (Error 3204, Line 7, Col 12, Line 7, Col 13, "If a multicase union type is a struct, then all union cases must have unique names. For example: 'type A = B of b: int | C of c: int'.")
+            (Error 3204, Line 7, Col 33, Line 7, Col 35, "If a multicase union type is a struct, then all union cases must have unique names. For example: 'type A = B of b: int | C of c: int'.")
+        ]
+        
+    [<Fact>]
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs:607]: This is such an exemplary testing that I will even happily forgive you the naming :D

**Review by psfinaki — APPROVED** [review]
Great job, again

File: tests/fsharp/typecheck/sigs/neg95.fs (hunk 7 above)
Comment (edgarfgp) [reply tests/fsharp/typecheck/sigs/neg95.fs:54]: I moved this test to the new ComponentTests/MultiCaseUnionStructTypes that I added as part of this PR

Comment (psfinaki) [reply tests/fsharp/typecheck/sigs/neg95.fs:54]: Ah right, sure

File: tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs (hunk 8 above)
Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/Conformance/UnionTypes/MultiCaseUnionStructTypes.fs:607]: Naming is hard you know  😅.

**Comment by T-Gro** [comment]
The error message was good before, but now it is not.
The condition is having about unique FIELD NAMES within the cases , with the special focus on making sure fields of the cases can have names (not everyone knows/uses named fields) and not to confuse them with the name of the case (which was a complaint about the existing message).

In the new version, the fact that this is about FIELDs has disappeared.

**Comment by psfinaki** [comment]
> If a multicase union type is a struct, then all union cases must have unique FIELD names. For example: "type A = B of b: int | C of c: int".

@T-Gro how about that? I wanted to remove the tautology in the previous version and ended up removing too much :D

**Comment by edgarfgp** [comment]
> The error message was good before, but now it is not. The condition is having about unique FIELD NAMES within the cases , with the special focus on making sure fields of the cases can have names (not everyone knows/uses named fields) and not to confuse them with the name of the case (which was a complaint about the existing message).
> 
> In the new version, the fact that this is about FIELDs has disappeared.

I think both are ok for me :) . Let me know what is the final error message and I will update it

**Comment by dsyme** [comment]
I'm happy to see this being improved!

**Comment by edgarfgp** [comment]
@T-Gro Thanks for merging this :)

**Comment by voronoipotato** [comment]
Thank you!!! this is huge as the previous error lacked an example demonstrating what it means in this context to name the cases, and how to do it.
